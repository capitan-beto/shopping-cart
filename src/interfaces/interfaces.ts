export interface Item {
    transactionID: string,
    id: number,
    title: string,
    price: number
}

export interface CartState {
    items: Item[]
}

export interface props {
    item: Item
}