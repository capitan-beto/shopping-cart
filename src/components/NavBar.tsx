import { Link } from "react-router-dom";
import styles from "../styles/header.module.css";
import useCart from "../hooks/useCart";

const NavBar = () => {
  const { cartCount } = useCart();

  return (
    <div>
        <nav>
          <div className="flex">
            <Link to="/" >
              <li>
                <p>Home</p>
              </li>
            </Link>
            <Link to="/Products" >
              <li>
                <p>Products</p>
              </li>
            </Link>
            <Link to="/Cart" >
              <li className="w-10">
                <p className="flex justify-center ml-2 m-0">
                  <span className="material-symbols-outlined py-2 text-xl">
                    shopping_cart
                  </span>
                  <span className={styles.lblCartCount}>
                    {cartCount}
                  </span>
                </p>
              </li>
            </Link>
          </div>
        </nav>
    </div>
  )
}

export default NavBar;