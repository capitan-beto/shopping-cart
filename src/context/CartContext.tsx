import { createContext } from "react";
import { CartState } from "../interfaces/interfaces";

export type CartContextProps = {
    cartState: CartState;
    removeFromCart: ( transactionID: string ) => void;
    addToCart: ( transactionID: string, id: number, title: string ) => void;
}

export const CartContext = createContext<CartContextProps>({} as CartContextProps);