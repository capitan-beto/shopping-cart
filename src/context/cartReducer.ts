import { CartState, Item } from "../interfaces/interfaces";

type CartAction = 
    | { type: "addToCart", payload: Item }
    | { type: "removeFromCart", payload: { id: number } };


export const cartReducer = ( state: CartState, action: CartAction ): CartState => {

    switch (action.type) {
        case "addToCart":
            return {
                ...state,
                items: [...state.items, action.payload]
            };

        case "removeFromCart":
            return {
                ...state,
                items: state.items.filter( item => {
                    item.id !== action.payload.id 
                })
            }
    
        default:
            return state;
    }
}