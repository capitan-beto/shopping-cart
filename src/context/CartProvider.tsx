import { useReducer } from "react";
import { CartState } from "../interfaces/interfaces";
import { CartContext } from "./CartContext";
import { cartReducer } from "./cartReducer";

const INITIAL_STATE: CartState = {
  items: []
}

export type propsType = {
  children: JSX.Element | JSX.Element[]
}

const CartProvider = ({ children }: propsType) => {

  const [cartState, dispatch] = useReducer( cartReducer, INITIAL_STATE );

  const removeFromCart = (transactionID: string) => {
    dispatch({ type: "removeFromCart", payload: { transactionID } })
  }

  const addToCart = (transactionID: string, id: number, title: string) => {
    dispatch({ type: "addToCart", payload: { transactionID, id, title} })
  }

  return (
    <CartContext.Provider value={{
      cartState,
      removeFromCart,
      addToCart
    }}>
        { children }
    </CartContext.Provider>
  )
}

export default CartProvider