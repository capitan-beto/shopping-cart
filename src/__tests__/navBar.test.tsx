import { render, screen } from "@testing-library/react";
import NavBar from "../components/NavBar";
import CartProvider from "../context/CartProvider";
import { BrowserRouter } from "react-router-dom";
import { CartContext, CartContextProps } from "../context/CartContext";
import { Item } from "../interfaces/interfaces";

const customRender = (ui: JSX.Element, {cartState, addToCart, removeFromCart}: CartContextProps) => {
    return render(
        <CartContext.Provider value={{cartState, addToCart, removeFromCart}}>
            {ui}
        </CartContext.Provider>,
        {wrapper: BrowserRouter}
    )
}

const props = (items: Item[]) : CartContextProps => {
    return {
        cartState: {
            items
        },
        removeFromCart: () => null,
        addToCart: () => null
    }
}

describe("Navbar tests", () => {
    it("cart icon should display initial cart items count", () => {
        render(
            <CartProvider>
                <NavBar />
            </CartProvider>,
            {wrapper: BrowserRouter}
        )

        expect(screen.getByTestId("cart-icon")).toMatchSnapshot();
    })

    it("cart icon should display correct number of items", () => {
        const testItems = [
            {title: "c3po", price: 1234, transactionID: "abv123", id: 1},
            {title: "r2d2", price: 4321, transactionID: "321vba", id: 2},
        ];

        customRender(<NavBar />, props(testItems));
        expect(screen.getByTestId("cart-icon")).toMatchSnapshot();
    })
})