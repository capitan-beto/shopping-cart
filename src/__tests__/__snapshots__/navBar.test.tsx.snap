// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Navbar tests cart icon should display correct number of items 1`] = `
<span
  class="lblCartCount"
  data-testid="cart-icon"
>
  2
</span>
`;

exports[`Navbar tests cart icon should display initial cart items count 1`] = `
<span
  class="lblCartCount"
  data-testid="cart-icon"
>
  0
</span>
`;
